### PR TITLE
feature/1365-ui-unified-search-filter

### PR DESCRIPTION
**Pull request by Nayana-R-Gowda**
Signed-off-by: NAYANAR <nayana.r7813@gmail.com>
#1365 

Files Modified
1. Backend - admin.py

Added tag_search query parameter to 5 partial endpoints:

/admin/servers/partial - Virtual Servers
/admin/tools/partial - Tools
/admin/prompts/partial - Prompts
/admin/resources/partial - Resources
/admin/a2a/partial - A2A Agents
Each endpoint filters across multiple database fields:

name, description, tags, owner_email
Plus entity-specific fields (e.g., url for tools, uri for resources)
2. Frontend - admin.js

Added performServerSideSearch() - triggers HTMX request with tag_search parameter
Added debouncedServerSearch() - 300ms debounce to avoid excessive requests
Updated clearSearch() - reloads data without filter
3. Templates - admin.html

Added oninput="debouncedServerSearch('entityType', this.value)" to all search inputs for reliable event handling
How It Works
User types in search box
After 300ms debounce, JavaScript sends HTMX GET request with ?tag_search=term&page=1
Server filters ALL records in database using SQL LIKE queries
Returns first page of matching results
Pagination automatically preserves the search term

### CODE REVIEW

**Comment by crivetimihai** [comment]
Thanks for working on server-side search, @Nayana-R-Gowda! Moving from client-side filtering (only the current page) to SQL-backed search across the full dataset is the right architectural call. The debounced HTMX approach and pagination integration look good.

A few things to address as you move this out of draft:

1. **28 test failures** — `func.lower()` calls are breaking existing test mocks. The mocked query objects likely don't support SQLAlchemy `func` operations. This needs investigation in the test fixtures.
2. **14 ESLint errors** — all appear auto-fixable (`npx eslint --fix`).
3. **Double event binding** — the HTML templates now have `oninput="debouncedServerSearch(...)"` inline AND `initializeSearchInputs()` adds `addEventListener("input", ...)` for the same inputs. Each keystroke will fire two search requests. Pick one approach, not both.
4. **Code repetition** — the search filter block is copy-pasted 5 times with minor variations. Consider extracting a helper like `apply_search_filter(query, model, search_term, fields)`.
5. **No new tests** — please add basic tests for the search functionality (filtered results, empty search, special characters).
6. **LIKE wildcard escaping** — user search terms containing `%` or `_` are not escaped, which could produce unexpected results. Minor UX issue.

The `clearSearch()` refactor using a lookup map is a nice cleanup!

File: mcpgateway/services/team_management_service.py [hunk 1]
```diff
@@ -38,6 +38,23 @@
...
+    Escapes % and _ characters to prevent unexpected wildcard matching
+    when users search for literal percent or underscore characters.
+
+    Args:
+        search_term: The raw search term from user input.
+
+    Returns:
+        str: Search term with % and _ characters escaped.
+    """
+    return search_term.replace("\\", "\\\\").replace("%", "\\%").replace("_", "\\_")
```
Comment (gcgoncalves) [inline mcpgateway/services/team_management_service.py]: This function is identical to the one on `admin.py`. They can be merged in a `utils.py` file and import with both.

File: mcpgateway/admin.py [hunk 2]
```diff
@@ -1503,6 +1561,7 @@ async def admin_servers_partial_html(
     include_inactive: bool = False,
     render: Optional[str] = Query(None),
     team_id: Optional[str] = Depends(_validated_team_id_param),
+    tag_search: Optional[str] = Query(None, description="Search across name, description, tags, owner, and team"),
```
Comment (gcgoncalves) [inline mcpgateway/admin.py]: `tag_search` seems like a search for tags. The `searchTerm` used in `admin.js` might be clearer, would you consider renaming this to `search_term`?

Also, as it's a user input variable, I recommend adding a length limit to it.

File: mcpgateway/admin.py [hunk 3]
```diff
@@ -462,6 +462,64 @@ async def wrapper(*args, request: Optional[Request] = None, **kwargs):
...
+            # JSON fields need to be cast to String
+            condition = func.lower(cast(field, String)).like(like_pattern, escape=escape_char)
+        elif is_nullable:
+            # Nullable fields need coalesce to handle NULLs
+            condition = func.lower(func.coalesce(field, "")).like(like_pattern, escape=escape_char)
+        else:
+            # Non-nullable string fields - direct comparison
+            # For ID fields (which are not lowered), use direct like
+            if hasattr(field, "type") and isinstance(field.type, String) and field.key == "id":
+                condition = field.like(like_pattern, escape=escape_char)
```
Comment (gcgoncalves) [inline mcpgateway/admin.py:521]: As the search term is lowered (line 498), I recommend either lowering the field as done on the else condition (removing the need for if altogether), or not lowering the search term when escaping it, an then lowering both the pattern and the field on the else clause.

File: mcpgateway/static/admin.js [hunk 4]
```diff
@@ -17225,12 +17202,147 @@ function clearSearch(entityType) {
...
+    });
+}
+
+/**
+ * Debounced server-side search to avoid excessive requests
+ * @param {string} entityType - The entity type
+ * @param {string} searchTerm - The search term
+ * @param {number} delay - Debounce delay in milliseconds (default: 300)
+ */
+function debouncedServerSearch(entityType, searchTerm, delay = 300) {
```
Comment (gcgoncalves) [inline mcpgateway/static/admin.js:17522]: Great choice debouncing the search!

File: mcpgateway/static/admin.js [hunk 5]
```diff
@@ -17163,58 +17163,35 @@ window.testSearchInit = function () {
...
-                filterGatewaysTable(""); // Clear the filter
+        // Map of entity types to their search input IDs
+        const inputMap = {
+            catalog: "catalog-search-input",
+            tools: "tools-search-input",
+            resources: "resources-search-input",
+            prompts: "prompts-search-input",
+            "a2a-agents": "a2a-agents-search-input",
+            gateways: "gateways-search-input",
+        };
```
Comment (gcgoncalves) [inline mcpgateway/static/admin.js:17381]: This refactor is really neat, good job!

File: mcpgateway/admin.py [hunk 6]
```diff
@@ -1581,6 +1641,22 @@ async def admin_servers_partial_html(
...
+            [
+                (DbServer.id, False, False),
+                (DbServer.name, False, False),
+                (DbServer.description, True, False),
+                (DbServer.tags, False, True),
+                (DbServer.owner_email, True, False),
+                (DbServer.federation_source, True, False),
+            ],
+        )
+        LOGGER.info(f"🔍 Servers search filter applied: {tag_search}")
```
Comment (gcgoncalves) [inline mcpgateway/admin.py]: The user input is being printed directly into log messages at INFO level. This has two issues:
 - Log injection risk: Crafted search terms with newlines or log format characters can corrupt log output or confuse log parsers.
 - Noise: Logging every search keystroke at INFO level in production (debounced at 300ms, but still high-frequency) will flood logs. Should be DEBUG at most.

Recommendation: Change to LOGGER.debug(...) and sanitise the search term (strip newlines and truncate to reasonable length)

File: mcpgateway/admin.py [hunk 7]
```diff
@@ -6625,6 +6709,23 @@ async def admin_tools_partial_html(
...
+                (DbTool.id, False, False),
+                (DbTool.original_name, False, False),
+                (DbTool.description, True, False),
+                (DbTool.tags, False, True),
+                (DbTool.owner_email, True, False),
+                (DbTool.url, True, False),
+                (DbTool.federation_source, True, False),
+            ],
+        )
+        LOGGER.info(f"🔍 Tools search filter applied: {tag_search}")
```
Comment (gcgoncalves) [inline mcpgateway/admin.py]: Same as [previous](https://github.com/IBM/mcp-context-forge/pull/2804/changes#r2792565329).

File: mcpgateway/admin.py [hunk 8]
```diff
@@ -7155,6 +7266,22 @@ async def admin_prompts_partial_html(
...
+            [
+                (DbPrompt.id, False, False),
+                (DbPrompt.name, False, False),
+                (DbPrompt.description, True, False),
+                (DbPrompt.tags, False, True),
+                (DbPrompt.owner_email, True, False),
+                (DbPrompt.federation_source, True, False),
+            ],
+        )
+        LOGGER.info(f"🔍 Prompts search filter applied: {tag_search}")
```
Comment (gcgoncalves) [inline mcpgateway/admin.py]: Same as [previous](https://github.com/IBM/mcp-context-forge/pull/2804/changes#r2792565329).

File: mcpgateway/admin.py [hunk 9]
```diff
@@ -7845,6 +7988,23 @@ async def admin_resources_partial_html(
...
+                (DbResource.id, False, False),
+                (DbResource.name, False, False),
+                (DbResource.description, True, False),
+                (DbResource.tags, False, True),
+                (DbResource.owner_email, True, False),
+                (DbResource.uri, False, False),
+                (DbResource.federation_source, True, False),
+            ],
+        )
+        LOGGER.info(f"🔍 Resources search filter applied: {tag_search}")
```
Comment (gcgoncalves) [inline mcpgateway/admin.py]: Same as [previous](https://github.com/IBM/mcp-context-forge/pull/2804/changes#r2792565329).

File: mcpgateway/admin.py [hunk 10]
```diff
@@ -8421,6 +8600,23 @@ async def admin_a2a_partial_html(
...
+                (DbA2AAgent.id, False, False),
+                (DbA2AAgent.name, False, False),
+                (DbA2AAgent.description, True, False),
+                (DbA2AAgent.tags, False, True),
+                (DbA2AAgent.owner_email, True, False),
+                (DbA2AAgent.endpoint_url, True, False),
+                (DbA2AAgent.federation_source, True, False),
+            ],
+        )
+        LOGGER.info(f"🔍 A2A Agents search filter applied: {tag_search}")
```
Comment (gcgoncalves) [inline mcpgateway/admin.py]: Same as [previous](https://github.com/IBM/mcp-context-forge/pull/2804/changes#r2792565329).

**Comment by Nayana-R-Gowda** [comment]
@gcgoncalves I’ve made the updates. Let me know if you need anything else.

**Review by gcgoncalves — APPROVED** [review]

**Comment by crivetimihai** [comment]
This is approved — just needs a rebase against main to resolve conflicts. Could you rebase and push? Then we can merge. Thanks, @Nayana-R-Gowda!

**Comment by crivetimihai** [comment]
Closing in favor of feat(ui): unified admin search with tags + global search #2868
